File: src/main/java/codingdojo/StoreEvent.java
package codingdojo;

/**
 * Ticket to In-store event, eg makeover, eyeshadow masterclass
 * or beauty product launch evening reception
 */
public class StoreEvent extends Product {
    protected Store location;

    public StoreEvent(String name, String id, Store location, Price price) {
        super(name, id, 0, price);
        setLocation(location);
    }

    public void setLocation(Store locationStore) {
        this.location = locationStore;
        location.addStoreEvent(this);
    }



    @Override
    public String toString() {
        return "StoreEvent{" +
                "id='" + id + '\'' +
                ", location=" + location +
                ", name='" + name + '\'' +
                ", price=" + price +
                '}';
    }

    @Override
    public boolean isEvent() {
        return true;
    }
}

File: TypeScript/src/StoreEvent.ts
import { Product } from "./Product";
import { Store } from "./Store";
import { Price } from "./Price";

/**
 * Ticket to In-store event, eg makeover, eyeshadow masterclass
 * or beauty product launch evening reception
 */
export class StoreEvent extends Product {

    private location: Store;

    constructor(name: string, id: string, location: Store, price: Price) {
        super(name, id, 0, price);

        this.location = location;
        this.location.addStoreEvent(this);
    }

    setLocation(store: Store) {
        this.location = store;
        this.location.addStoreEvent(this);
    }

    toString(): string {
        return "StoreEvent{" + this.name + "}";
    }

    isEvent(): boolean {
        return true;
    }

}
